python/dask_cudf/dask_cudf/sorting.py

```diff
@@ -161,9 +168,16 @@ def rearrange_by_hash(
     }
 
     shuffle_group_token = "shuffle-group-" + token
+    shuffle_split_token = "shuffle-split-" + token
+    k_list = tuple(range(k))
+
     for stage in range(1, stages + 1):
-        group = {  # Convert partition into dict of dataframe pieces
-            (shuffle_group_token, stage, inp): (
+        group = {}
+        split = {}
+        combine = {}
+        for inp in inputs:
```

Comment:
You may be able to fuse handling `end` at line 221 as well as `start` at line 163 into the loop where we generate `inputs` at line 156.